### PR TITLE
Add a cohort creation command

### PR DESCRIPTION
**Pull request by maxcell**
### Usage
The command usage would be something like `!create_cohort work-in-public` and then it will create the channel and the role. It will also apply the locked access for the created channel.

image

### Technical Notes
In order to support adding more commands for the future, I converted the code to begin using the command framework from serenity. A lot of the patterns were adapted from their examples (see [command framework](https://github.com/serenity-rs/serenity/blob/current/examples/e05_command_framework/src/main.rs) and [sample_bot_structure](https://github.com/serenity-rs/serenity/tree/current/examples/e07_sample_bot_structure)).

 There is a very light check for duplicate role names but nothing for channels. If we feel it is important before merging, we can make sure to codify any other rules!

### CODE REVIEW

File: rust-bot/src/main.rs [hunk 1]
```diff
@@ -1,41 +1,108 @@
...
+    return Ok(());
+}
+
+// This command provisions out a channel with permissions
+// to read/write for users with the corresponding role. It
+// will also make the channel read-only for other users.
+#[command]
+fn create_cohort(ctx: &mut Context, msg: &Message, mut args: Args) -> CommandResult {
+    msg.channel_id
+        .say(&ctx.http, "Spinning up Adventure Club...")?;
```
Comment (ChristopherBiscardi) [inline rust-bot/src/main.rs:41]: love this :D

File: rust-bot/src/main.rs [hunk 2]
```diff
@@ -1,41 +1,108 @@
...
-    // Set a handler to be called on the `ready` event. This is called when a
-    // shard is booted, and a READY payload is sent by Discord. This payload
-    // contains data like the current user's guild Ids, current user data,
-    // private channels, and more.
-    //
-    // In this case, just print what the current user's username is.
-    fn ready(&self, _: Context, ready: Ready) {
-        println!("{} is connected!", ready.user.name);
+#[command]
+fn ping(ctx: &mut Context, msg: &Message) -> CommandResult {
```
Comment (ChristopherBiscardi) [inline rust-bot/src/main.rs:28]: so much nicer with the subsystem commands :D

File: rust-bot/src/main.rs [hunk 3]
```diff
@@ -1,41 +1,108 @@
...
+    let role = create_role(ctx, msg, &cohort_name);
+    let channel = create_channel(ctx, msg, &channel_name, role);
+
+    match channel {
+        Ok(channel) => {
+            let reply_msg = format!(
+                "Successfully created {}! Feel free to add users.",
+                channel.name
+            );
+            msg.reply(&ctx.http, reply_msg).unwrap();
```
Comment (ChristopherBiscardi) [inline rust-bot/src/main.rs:53]: these unwraps should log data into honeycomb when @mdarrik gets the implementation up

File: rust-bot/src/main.rs [hunk 4]
```diff
@@ -1,41 +1,108 @@
...
+
+// Permissions for chatting happen here
+fn create_channel(
+    ctx: &mut Context,
+    msg: &Message,
+    channel_name: &str,
+    role: Role,
+) -> Result<GuildChannel, Error> {
+    let role_id = role.id;
+    let everyone_id = get_everyone_role(ctx, msg).unwrap().id;
```
Comment (ChristopherBiscardi) [inline rust-bot/src/main.rs:89]: maybe use `role_by_name` here?

File: rust-bot/src/main.rs [hunk 5]
```diff
@@ -1,41 +1,108 @@
...
+    let permission_set = mute_users_without_role_permset(role_id, everyone_id);
+
+    let guild = msg.guild(&ctx.cache).unwrap();
+    let new_channel = guild.read().create_channel(&ctx.http, |c| {
+        c.name(channel_name)
+            .kind(ChannelType::Text)
+            .permissions(permission_set)
```
Comment (ChristopherBiscardi) [inline rust-bot/src/main.rs:96]: Add mods to permissions set, with the ability to do anything that is denied to `@everyone`

File: rust-bot/src/main.rs [hunk 6]
```diff
@@ -53,3 +130,54 @@ fn main() {
         println!("Client error: {:?}", why);
     }
 }
+
+#[check]
+#[name = "Mod"]
+fn mod_check(ctx: &mut Context, msg: &Message, _: &mut Args, _: &CommandOptions) -> CheckResult {
+    if let Some(member) = msg.member(&ctx.cache) {
+        if let Ok(permissions) = member.permissions(&ctx.cache) {
+            return permissions.administrator().into();
```
Comment (ChristopherBiscardi) [inline rust-bot/src/main.rs:139]: should be `Moderator` role, instead of administrator.

File: rust-bot/src/main.rs [hunk 7]
```diff
@@ -1,41 +1,108 @@
...
+            .kind(ChannelType::Text)
+            .permissions(permission_set)
+    });
+
+    new_channel
 }
 
+struct Handler;
+
+impl EventHandler for Handler {}
```
Comment (ChristopherBiscardi) [inline rust-bot/src/main.rs:104]: What is the empty block here doing?

File: rust-bot/src/main.rs [hunk 8]
```diff
@@ -53,3 +130,54 @@ fn main() {
...
+        },
+        PermissionOverwrite {
+            allow: Permissions::empty(),
+            deny: remove_messaging,
+            kind: PermissionOverwriteType::Role(everyone_role_id),
+        },
+    ]
+}
+
+fn get_everyone_role(ctx: &mut Context, msg: &Message) -> Option<Role> {
```
Comment (ChristopherBiscardi) [inline rust-bot/src/main.rs:175]: `role_by_name` instead?

**Review by ChristopherBiscardi — APPROVED** [review]

**Comment by maxcell** [comment]
todo for you future prince: move channel into group

**Pull request merged**